frontend/controller/serviceworkers/push.js

```diff
@@ -7,28 +7,83 @@ import { NOTIFICATION_TYPE, PUBSUB_RECONNECTION_SUCCEEDED, PUSH_SERVER_ACTION_TY
+      'base64'
+    )
+    return v
+  }
+  return new Uint8Array(v)
+}
+
+const bufferEq = (a?: ArrayBuffer | Uint8Array, b?: ArrayBuffer | Uint8Array) => {
+  // eslint-disable-next-line eqeqeq
+  if (a == null || b == null) return a == b
+
+  const ab = strOrBufToBuf(a)
+  const bb = strOrBufToBuf(b)
+
+  if (ab.byteLength !== bb.byteLength) return false
+
+  for (let i = ab.byteLength - 1; i >= 0; i--) {
+    if (ab[i] !== bb[i]) return false
+  }
+
+  return true
+}
+
 export default (sbp('sbp/selectors/register', {
   'push/getSubscriptionOptions': (() => {
     let cachedVapidInformation
+
+    sbp('okTurtles.events/on', REQUEST_TYPE.PUSH_ACTION, ({ data }) => {
+      if (data.type !== PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY) return
+      const oldKey = cachedVapidInformation?.[1].applicationServerKey
+      cachedVapidInformation = [performance.now(), {
+        userVisibleOnly: true,
+        applicationServerKey: data.data
+      }]
+      if (oldKey === data.data) return
+      (async () => {
+        const subscription = await self.registration.pushManager.getSubscription()
+        if (!subscription) return
+        if (bufferEq(subscription.options.applicationServerKey, data.data)) return
+        console.warn('VAPID server key changed; removing existing subscription and setting up a new one', {
+          oldApplicationServerPublicKey: subscription.options.applicationServerKey && Buffer.from(subscription.options.applicationServerKey).toString('base64'),
+          newApplicationServerPublicKey: data.data
+        })
+        await subscription.unsubscribe()
+        // return b/c device settings not loaded
+        if (!sbp('chelonia/rootState').loggedIn || sbp('sw/deviceSettings/get', DEVICE_SETTINGS.DISABLE_NOTIFICATIONS)) return
+        const newSubscription = await self.registration.pushManager.subscribe(cachedVapidInformation[1])
+        await sbp('push/reportExistingSubscription', newSubscription?.toJSON(), newSubscription?.options.applicationServerKey)
+      })()
+    })
+
     return () => {
       if (
         cachedVapidInformation &&
         // Cache the VAPID information for one hour. The server public
         // information should change very infrequently, if it changes at all.
         (performance.now() - cachedVapidInformation[0]) < 3600e3
       ) {
-        return cachedVapidInformation[1]
+        return Promise.resolve(cachedVapidInformation[1])
       }
 
       const result = new Promise((resolve, reject) => {
         const handler = ({ data }) => {
           if (data.type !== PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY) return
           sbp('okTurtles.events/off', REQUEST_TYPE.PUSH_ACTION, handler)
           clearTimeout(timeoutId)
-          resolve({
-            userVisibleOnly: true,
-            applicationServerKey: data.data
-          })
+          // Set in the handler above
+          resolve(cachedVapidInformation[1])
```

Comment:
This code seems somewhat brittle to me. It says `Set in the handler above`, and that is technically true... and it is technically true that that code will also run first because that handler was registered first.... but if the order were different this could resolve with `undefined` or even throw an exception (`undefined[1]`)

Reply:
But the order is not different..? We need to wait on the server response being returned, and we already have the handler above to handle that message, and so we don't need to repeat the same logic. I'm not sure how to keep the same semantics in a simpler way.

Reply:
Maybe a comment acknowledging this reality would be helpful to readers who aren't familiar with how `okTurtles.events` works, to let them know that this code only works because of the order in which handlers were registered as `okTurtles.events` loops through event handlers in that order when emitting events.

Reply:
Do you have a suggestion for improving this otherwise to make it less brittle while keeping it simple too? I noticed, for example, that it would break if `debounce` is used as suggested in https://github.com/okTurtles/group-income/pull/2704#discussion_r2040149550